src/audio/kpb.c

```diff
@@ -527,17 +536,105 @@ static void kpb_init_draining(struct comp_data *kpb, struct kpb_client *cli)
 	} else if (!is_sink_ready) {
 		trace_kpb_error("kpb_init_draining() error: "
 				"sink not ready for draining");
+
+	} else if (!kpb_have_enough_history_data(buff, history_depth)) {
+		trace_kpb_error("kpb_init_draining() error: "
+				"not enough data in history buffer");
 	} else {
+		/* draining accepted, find proper buffer to start reading */
+		while (buff->next && buff->next != first_buff) {
+			if (buff->state == KPB_BUFFER_FREE)
+				break;
+			else
+				buff = buff->next;
+		}
+
+		while (buffered < history_depth) {
+			if (buff->state == KPB_BUFFER_FULL) {
+				buffered += (int)buff->end_addr -
+					    (int)buff->sta_addr;
+			} else {
+				buffered += (int)buff->w_ptr -
+					    (int)buff->sta_addr;
+			}
+
+			buff = buff->next;
+		}
+
+		read_offset = buffered - history_depth;
+
+		if (read_offset > 0)
+			buff->r_ptr = buff->sta_addr + read_offset;
+		else
+			buff->r_ptr = buff->sta_addr;
+
 		/* add one-time draining task into the scheduler */
-		schedule_task(&kpb->draining_task, 0, 0);
+		kpb->draining_task_data.sink = cli->sink;
+		kpb->draining_task_data.history_buffer = buff;
+
+		schedule_task_idle(&kpb->draining_task, 100);
 	}
 }
 
-static void kpb_draining_task(void *arg)
+static void draining_task(void *arg)
 {
-	/* TODO: while loop drainning history buffer accoriding to
-	 * clients request
-	 */
+	struct dd *draining_data = (struct dd *)arg;
+	struct comp_buffer *sink = draining_data->sink;
+	struct hb *buff = draining_data->history_buffer;
+	size_t size_to_read = (int)buff->end_addr - (int)buff->r_ptr;
```

Comment:
please check your casting, in many other places too.

Reply:
In general, instead of int there is intptr_t for such operations